core/notifee.go

```diff
@@ -0,0 +1,81 @@
+// +build !js
+
+package core
+
+import (
+	"time"
+
+	p2pnet "github.com/libp2p/go-libp2p-net"
+	ma "github.com/multiformats/go-multiaddr"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	// pubsubProtocolTag is the tag used for peers who speak our pubsub protocol.
+	pubsubProtocolTag = "pubsub-protocol"
+	// pubsubProtocolScore is the score to add to peers who speak our pubsub
+	// protocol.
+	pubsubProtocolScore = 10
+)
+
+// notifee receives notifications for network-related events.
+type notifee struct {
+	node *Node
+}
+
+var _ p2pnet.Notifiee = &notifee{}
+
+// Listen is called when network starts listening on an addr
+func (n *notifee) Listen(p2pnet.Network, ma.Multiaddr) {}
+
+// ListenClose is called when network stops listening on an addr
+func (n *notifee) ListenClose(p2pnet.Network, ma.Multiaddr) {}
+
+// Connected is called when a connection opened
+func (n *notifee) Connected(network p2pnet.Network, conn p2pnet.Conn) {
+	log.WithFields(map[string]interface{}{
+		"peerID":       conn.RemotePeer(),
+		"multiaddress": conn.RemoteMultiaddr(),
+	}).Trace("connected to peer")
+}
+
+// Disconnected is called when a connection closed
+func (n *notifee) Disconnected(network p2pnet.Network, conn p2pnet.Conn) {
+	log.WithFields(map[string]interface{}{
+		"peerID":       conn.RemotePeer(),
+		"multiaddress": conn.RemoteMultiaddr(),
+	}).Trace("disconnected from peer")
+}
+
+// OpenedStream is called when a stream opened
+func (n *notifee) OpenedStream(network p2pnet.Network, stream p2pnet.Stream) {
+	go func() {
+		// HACK(albrow): When the stream is initially opened, the protocol is not set.
+		// For now, we have to manually poll until it is set.
```

Comment:
Is there a PR or issue you could add a link to?

Reply:
Just added a link to this issue: https://github.com/libp2p/go-libp2p/issues/467. I'm not 100% sure if it will help us avoid this hack; it depends on how exactly the event bus is implemented.